dppl/oneapi_interface.pyx

```diff
@@ -101,37 +95,70 @@ cdef class DpplRuntime:
-        '''
-        cdef void* queue_ptr = NULL;
-        self.rt.getCurrentQueue(&queue_ptr);
-        return PyCapsule_New(queue_ptr, NULL, &delete_queue)
-
-#    def set_global_queue (self, device_ty, device_id):
-#        if device_ty == device_type.gpu:
-#            self.rt.setGlobalContextWithGPU(device_id)
-#        elif device_ty == device_type.cpu:
-#            self.rt.setGlobalContextWithCPU(device_id)
-#        else:
-#            e = UnsupportedDeviceTypeError("Device can only be cpu or gpu")
-#            raise e
-
     def dump (self):
         ''' Prints information about the Runtime object.
         '''
         return self.rt.dump()
 
-    def dump_queue (self, queue_cap):
+    def dump_queue_info (self, queue_cap):
         ''' Prints information about the SYCL queue object.
         '''
         if PyCapsule_IsValid(queue_cap, NULL):
-            self.rt.dump_queue(PyCapsule_GetPointer(queue_cap, NULL))
+            return self.rt.dump_queue(PyCapsule_GetPointer(queue_cap, NULL))
         else:
             raise ValueError("Expected a PyCapsule encapsulating a SYCL queue")
 
-# Global runtime object
-runtime = DpplRuntime()
+    def get_current_queue (self):
+        ''' Returns the activated SYCL queue as a PyCapsule.
+        '''
+        cdef void* queue_ptr = NULL;
+        self.rt.getCurrentQueue(&queue_ptr);
+        return PyCapsule_New(queue_ptr, NULL, &delete_queue)
+
+    def get_num_platforms (self):
+        ''' Returns the number of available SYCL/OpenCL platforms.
+        '''
+        cdef size_t num_platforms = 0
+        self.rt.getNumPlatforms(&num_platforms)
+        return num_platforms
+
+    def set_default_queue (self, device_ty, device_id):
+        if device_ty == device_type.gpu:
+            self.rt.resetGlobalQueue(_device_type._GPU, device_id)
+        elif device_ty == device_type.cpu:
+            self.rt.resetGlobalQueue(_device_type._CPU, device_id)
+        else:
+            e = UnsupportedDeviceTypeError("Device can only be cpu or gpu")
+            raise e
+
+    def is_in_dppl_ctxt (self):
+        cdef size_t num = 0
+        self.rt.number_of_activated_queues(num)
+        if num:
+            return True
+        else:
+            return False
+
+
+# thread-local storage
+from threading import local as threading_local
+
+# Initialize a thread local instance of _DpplRuntime
+_tls = threading_local()
+_tls._runtime = _DpplRuntime()
+
+
+################################################################################
+#--------------------------------- Public API ---------------------------------#
+################################################################################
+
+
+dump              = _tls._runtime.dump
+dump_queue_info   = _tls._runtime.dump_queue_info
+get_current_queue = _tls._runtime.get_current_queue
+get_num_platforms = _tls._runtime.get_num_platforms
+set_default_queue = _tls._runtime.set_default_queue
+is_in_dppl_ctxt   = _tls._runtime.is_in_dppl_ctxt
```

Comment:
Why is the entire runtime thread-local? The only thing which is thread-local is the stack of contexts. Making shared data thread-local is calling for trouble.
I think the runtime class should encapsulate the thread-local business and return the thread-local values in its member functions like get_current_queue. Maybe this needs to be done in C++ (e.g. your std::deque be thread-local).